core/src/main/java/com/linecorp/armeria/common/QueryStringDecoder.java

```diff
@@ -0,0 +1,195 @@
+import java.nio.charset.StandardCharsets;
+
+import javax.annotation.Nullable;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import io.netty.util.CharsetUtil;
+
+final class QueryStringDecoder {
+
+    // Forked from netty-4.1.43.
+    // https://github.com/netty/netty/blob/7d6d953153697bd66c3b01ca8ec73c4494a81788/codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java
+
+    @SuppressWarnings("checkstyle:FallThrough")
+    static QueryParams decodeParams(@Nullable String s, int paramsLimit, boolean semicolonIsSeparator) {
+        if (s == null) {
+            return QueryParams.of();
+        }
+
+        final int len = s.length();
+
+        // Skip the leading question marks.
+        int from = -1;
+        for (int i = 0; i < len; i++) {
+            if (s.charAt(i) != '?') {
+                from = i;
+                break;
+            }
+        }
+
+        // Return an empty parameters if the string is empty or full of question marks.
+        if (from < 0) {
+            return QueryParams.of();
+        }
+
+        final QueryParamsBuilder params = QueryParams.builder();
+        int nameStart = from;
+        int valueStart = -1;
+        int i;
+        loop:
+        for (i = from; i < len; i++) {
+            switch (s.charAt(i)) {
+                case '=':
+                    if (nameStart == i) {
+                        nameStart = i + 1;
+                    } else if (valueStart < nameStart) {
+                        valueStart = i + 1;
+                    }
+                    break;
+                case ';':
+                    if (!semicolonIsSeparator) {
+                        continue;
+                    }
+                    // fall-through
+                case '&':
+                    if (addParam(s, nameStart, valueStart, i, params)) {
+                        paramsLimit--;
+                        if (paramsLimit == 0) {
+                            return params.build();
+                        }
+                    }
+                    nameStart = i + 1;
+                    break;
+                case '#':
+                    break loop;
+                default:
+                    // continue
+            }
+        }
+        addParam(s, nameStart, valueStart, i, params);
+        return params.build();
+    }
+
+    private static boolean addParam(String s, int nameStart, int valueStart, int valueEnd,
+                                    QueryParamsBuilder params) {
+        if (nameStart >= valueEnd) {
+            return false;
+        }
+        if (valueStart <= nameStart) {
+            valueStart = valueEnd + 1;
+        }
+        final String name = decodeComponent(s, nameStart, valueStart - 1);
+        final String value = decodeComponent(s, valueStart, valueEnd);
+        params.add(name, value);
+        return true;
+    }
+
+    @VisibleForTesting
+    static String decodeComponent(String s, int from, int toExcluded) {
```

Comment:
Can we benchmark this against JDK's `URLDecoder`? Looking at the implementation in Java 13, I don't see any reason for it to be any slower than this code.

Reply:
Realized the biggest performance advantage is probably not needing to copy into `substring`. Lot's of code for that 😅  If keeping it, I would probably replace the `ByteBuffer` and `CharBuffer` with `byte[]` and `new String(bytes, int, int, charset)`, `ByteBuffer` should basically always be at least as slow as native array.

Reply:
By the way I found https://github.com/netty/netty/pull/6781/files which switched from `byte[]` to `ByteBuffer`. I'm not sure why - anyways we can consider reworking it here or doing it in Netty and backporting if it ends up being better.

Reply:
Sent https://github.com/netty/netty/pull/9886 - seeing noticably better performance with an array

Reply:
And also noticed... we could probably beat Netty using our `ThreadLocalByteArray` :)

Reply:
We could optimize this further given that we're always decoding UTF-8.